core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/common/index/EqualsAndComparisonIndexer.java

```diff
@@ -0,0 +1,122 @@
+        NavigableMap<Object, Map<Tuple_, Value_>> comparisonMap = equalsMap.get(equalsIndexKey);
+        if (comparisonMap == null) {
+            throw new IllegalStateException("Impossible state: the tuple (" + tuple
+                    + ") with indexProperties (" + Arrays.toString(indexProperties)
+                    + ") doesn't exist in the indexer.");
+        }
+        Map<Tuple_, Value_> tupleMap = comparisonMap.get(comparisonIndexProperty);
+        if (tupleMap == null) {
+            throw new IllegalStateException("Impossible state: the tuple (" + tuple
+                    + ") with indexProperties (" + Arrays.toString(indexProperties)
+                    + ") doesn't exist in the indexer.");
+        }
+        Value_ value = tupleMap.remove(tuple);
+        if (value == null) {
+            throw new IllegalStateException("Impossible state: the tuple (" + tuple
+                    + ") with indexProperties (" + Arrays.toString(indexProperties)
+                    + ") doesn't exist in the indexer.");
+        }
+        if (tupleMap.isEmpty()) {
+            comparisonMap.remove(comparisonIndexProperty);
+            if (comparisonMap.isEmpty()) {
+                equalsMap.remove(equalsIndexKey);
+            }
+        }
+        return value;
+    }
+
+    @Override
+    public Map<Tuple_, Value_> get(Object[] indexProperties) {
+        IndexerKey equalsIndexKey = new IndexerKey(Arrays.copyOfRange(indexProperties, 0, indexProperties.length - 1));
+        Object comparisonIndexProperty = indexProperties[indexProperties.length - 1];
+        NavigableMap<Object, Map<Tuple_, Value_>> comparisonMap = equalsMap.get(equalsIndexKey);
+        if (comparisonMap == null) {
+            return Collections.emptyMap();
+        }
+        NavigableMap<Object, Map<Tuple_, Value_>> selectedComparisonMap;
+        switch (comparisonJoinerType) {
+            case LESS_THAN:
+                selectedComparisonMap = comparisonMap.headMap(comparisonIndexProperty, false);
+                break;
+            case LESS_THAN_OR_EQUAL:
+                selectedComparisonMap = comparisonMap.headMap(comparisonIndexProperty, true);
+                break;
+            case GREATER_THAN:
+                selectedComparisonMap = comparisonMap.tailMap(comparisonIndexProperty, false);
+                break;
+            case GREATER_THAN_OR_EQUAL:
+                selectedComparisonMap = comparisonMap.tailMap(comparisonIndexProperty, true);
+                break;
+            default:
+                throw new IllegalStateException("Impossible state: the comparisonJoinerType (" + comparisonJoinerType
+                        + ") is not one of the 4 comparison types.");
+        }
+        return selectedComparisonMap.values().stream()
+                .flatMap(map -> map.entrySet().stream())
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```

Comment:
As this is on the hot path for indexing, I would suggest unrolling this into traditional loops for performance.
Streams, and in this case nested streams, are going to be unnecessarily slow here.

Reply:
Good point. Separate issue, added it as a task here: https://issues.redhat.com/browse/PLANNER-2675